src/Common/src/CoreLib/System/IO/PathInternal.cs

```diff
@@ -117,24 +126,51 @@ internal static bool AreRootsEqual(string first, string second, StringComparison
         internal static string RemoveRelativeSegments(string path, int rootLength)
         {
             Debug.Assert(rootLength > 0);
-            bool flippedSeparator = false;
 
+            Span<char> initialBuffer = stackalloc char[MaxStackPath];
+            ValueStringBuilder sb = new ValueStringBuilder(initialBuffer);
+
+            bool pathNotChanged = RemoveRelativeSegmentsHelper(path.AsSpan(), rootLength, ref sb);
+
+            if (sb.Length >= rootLength && !pathNotChanged)
+                return sb.ToString();
+
+            // If we haven't changed the source path, return the original
+            return pathNotChanged ? path : path.Substring(0, rootLength);
```

Comment:
Seems like this would be simpler to understand if pathNotChanged was checked first:
```C#
return
    pathNotChanged ? path :
    sb.Length >= rootLength ? sb.ToString() :
    path.Substring(0, rootLength);
```

Reply:
I'm a little confused as to why we need to check that the length is greater than the root length. Shouldn't that never happen?

Reply:
https://github.com/dotnet/corefx/pull/28835/files#diff-200474c05bdb868e1d141aeeb98cbf18R138